modules/internal/ChapelLocale.chpl

```diff
@@ -247,6 +247,35 @@ module ChapelLocale {
       return "";
     }
 
+    //
+    // Support for different types of memory:
+    // large, random-access (low latency), and streaming (high bandwidth)
+    // Inspired by madvise()'s support for RANDOM and SEQUENTIAL access
+    //
+    pragma "no doc"
+    proc defaultMemory() : locale {
+      _throwPVFCError();
+      return this;
+    }
+
+    pragma "no doc"
+    proc largeMemory() : locale {
+      _throwPVFCError();
+      return this;
+    }
+
+    pragma "no doc"
+    proc randomMemory() : locale {
+      _throwPVFCError();
+      return this;
+    }
+
+    pragma "no doc"
+    proc streamingMemory() : locale {
+      _throwPVFCError();
+      return this;
+    }
+
```

Comment:
The `madvise(2)` differentiation between random and streaming access does make sense to me.  But that's intended to tell the kernel something about the program's behavior with respect to the given memory segment.  Locale models, on the other hand, are about architecture rather than program behavior and I think we should be describing the characteristics of the memory itself rather than the program access behavior that would match up with it.  So I'd prefer to see the random vs. streaming differentiation expressed as (optimized for) latency vs. bandwidth.  I'd be interested in what others think about this, though.  Maybe our notion of a "locale model" needs to be expanded to include how programs expect to interact with the architecture, rather than just being about the architecture as such.

Another topic: this model allows a program to express at most one memory characteristic and get either a matching memory sublocale or, failing that, a fallback sublocale chosen by the locale model.  Some time ago we had discussed, and I thought informally settled on, a model in which programmers would somehow be able to express an ordered set of characteristics and the returned sublocale would be the locale model's best attempt to match all of those.  For example you could say something like `selectMemory(large, latency)` to say that you wanted a memory that was both large and latency-optimized but that the size was more important to you than optimization for access pattern.  Or maybe you'd even be able to say `selectMemory(sizeMin=1**30, bandwidth)` where sizeMin was a required characteristic, to say that you wanted a bandwidth-optimized memory but it had to be at least 1gb in size.  This isn't actually so important when all we have to choose from is size and latency vs. bandwidth, but the day may not be far off when memories also have volatility and/or resilience characteristics, and it could be very important to be able to express that you want a large resilient memory for your long-lasting in-memory database. It looks hard to achieve this with characteristic-specific selector functions as you've implemented here.